nearcore/src/state_sync.rs

```diff
@@ -194,79 +233,110 @@ async fn state_sync_dump(
+                                let mut parts_to_dump = parts_not_dumped.clone();
+                                let cnt_parts_to_dump: u64 =
+                                    parts_to_dump.len().try_into().unwrap();
+                                let mut cnt_parts_dumped = num_parts - cnt_parts_to_dump;
+                                let timer = Instant::now();
+                                while timer.elapsed().as_secs() <= 60 && !parts_to_dump.is_empty() {
+                                    let _timer = metrics::STATE_SYNC_DUMP_ITERATION_ELAPSED
+                                        .with_label_values(&[&shard_id.to_string()])
+                                        .start_timer();
+
+                                    let (part_id, selected_idx) =
+                                        select_random_part_id_with_index(&parts_to_dump);
+
+                                    let state_part = match obtain_and_store_state_part(
+                                        runtime.as_ref(),
+                                        shard_id,
+                                        sync_hash,
+                                        &sync_prev_hash,
+                                        &state_root,
+                                        part_id,
+                                        num_parts,
+                                        &chain,
+                                    ) {
+                                        Ok(state_part) => state_part,
+                                        Err(_) => {
+                                            break;
+                                        }
+                                    };
+                                    let location = external_storage_location(
+                                        &chain_id,
+                                        &epoch_id,
+                                        epoch_height,
+                                        shard_id,
+                                        part_id,
+                                        num_parts,
+                                    );
+                                    if let Err(_) = external
+                                        .put_state_part(&state_part, shard_id, &location)
+                                        .await
+                                    {
+                                        // no need to break if there's an error, we should keep dumping other parts.
+                                        // reason is we are dumping random selected parts, so it's fine if we are not able to finish all of them
+                                        continue;
+                                    }
+
+                                    // remove the dumped part from parts_to_dump so that we draw without replacement
+                                    parts_to_dump.swap_remove(selected_idx);
+                                    cnt_parts_dumped += 1;
```

Comment:
Would be nice to update Prometheus metrics that one more part was dumped and it X more bytes were uploaded.